### PR TITLE
feat(react): add Listbox component

### PR DESCRIPTION
**Pull request by scurker**
see #1088 

`Listbox` will be one of the driving behaviors behind `Combobox` to provide that set of functionality for the combobox list. Eventually, we can also refactor `OptionMenu` to also utilize this component as well to help consolidate these types of behaviors within a single component.

### CODE REVIEW

File: docs/pages/components/Listbox.mdx [hunk 1]
```diff
@@ -0,0 +1,212 @@
+---
+title: Listbox
+description: An unstyled component to provide a keyboard navigable list of options.
+source: https://github.com/dequelabs/cauldron/tree/develop/packages/react/src/components/Listbox/Listbox.tsx
```
Comment (anastasialanz) [inline docs/pages/components/Listbox.mdx:4]:
```suggestion
description: An unstyled component to provide a keyboard-navigable list of options.
source: https://github.com/dequelabs/cauldron/tree/develop/packages/react/src/components/Listbox/Listbox.tsx
```

File: docs/pages/components/Listbox.mdx [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+]} />
+
+### ListboxGroup
+
+<ComponentProps className={true} refType="HTMLUListElement" props={[
+  {
+    name: 'label',
+    required: true,
+    type: ['string', 'number'],
+    description: 'Value to be applied to the listbox option. When omitted, value will default to the text content of the option.'
```
Comment (dequejosie) [inline docs/pages/components/Listbox.mdx:204]: Is this `description` correct? It seems that this prop is required, how is it omitted?

File: packages/styles/forms.css [hunk 3]
```diff
@@ -58,7 +60,6 @@ textarea,
 [role='menuitemcheckbox'],
 [role='menuitemradio'],
 [role='textbox'],
-[aria-haspopup='listbox'],
```
Comment (dequejosie) [inline packages/styles/forms.css:61]: What were these styles for?

File: docs/pages/components/Listbox.mdx (hunk 2 above)
Comment (scurker) [reply docs/pages/components/Listbox.mdx]: Whoops! Bad copy pasta, fixed.

File: packages/styles/forms.css (hunk 3 above)
Comment (scurker) [reply packages/styles/forms.css:61]: These styles were overly aggressive and getting applied incorrectly. The prior styles were assuming that anything that had `[aria-haspopup='listbox']` should be styled like an input - which is incorrect.

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    target: T;
+    value: ListboxValue;
+  }) => void;
+  onSelectionChange?: ({
+    value
+  }: {
+    previousValue: ListboxValue;
+    value: ListboxValue;
+  }) => void;
+  onActiveChange?: (option: ListboxOption) => void;
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/Listbox.tsx:33]: Seems this prop is not documented.

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    }, [activeOption]);
+
+    const handleSelect = useCallback(
+      (option: ListboxOption) => {
+        setActiveOption(option);
+        onSelect?.({ target: option.element, value: option.value as string });
+        // istanbul ignore else
+        if (!isControlled) {
+          setSelectedOption(option);
+          onSelectionChange?.({
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/Listbox.tsx:98]: It seems that `onSelectionChange` only applies to uncontrolled `listbox`. It makes sense, cause controlled `listbox` has `onSelect`, and uncontrolled `listbox` has to rely on `onSelectionChange` if the user wants to do something whenever the selection changes. it could be confusing if a controlled listbox user accidentally passes in `onSelectionChange` and it does not do anything.

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        }
+      },
+      [isControlled, selectedOption]
+    );
+
+    const handleKeyDown = useCallback(
+      (event: React.KeyboardEvent<HTMLElement>) => {
+        onKeyDown?.(event);
+
+        if (!keys.includes(event.key)) return;
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/Listbox.tsx:111]: nit:
```suggestion
        if (!keys.includes(event.key)) { 
          return
        };
```

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        event.preventDefault();
+        const enabledOptions = options.filter(
+          (option) => !isDisabledOption(option)
+        );
+
+        // istanbul ignore next
+        if (!enabledOptions.length) return;
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/Listbox.tsx:118]: The same.

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+          ({ element }) => element === currentOption.element
+        );
+        const allowCyclicalNavigation = navigation === 'cycle';
+
+        switch (event.key) {
+          case up:
+            const previousOption =
+              currentIndex === 0 && allowCyclicalNavigation
+                ? lastOption
+                : enabledOptions[Math.max(currentIndex - 1, 0)];
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/Listbox.tsx:140]: this always returns 0, right?

File: packages/react/src/components/Listbox/ListboxContext.tsx [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+} & ListboxContext<T>;
+
+/* istanbul ignore next */
+const ListboxContext = createContext({
+  options: [],
+  active: null,
+  selected: null,
+  setOptions: () => null,
+  onSelect: () => null
+});
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/ListboxContext.tsx:29]: Why is typescript not happy here?

File: packages/react/src/components/Listbox/ListboxGroup.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+interface ListboxGroupProps extends React.HTMLAttributes<HTMLElement> {
+  as?: React.ElementType | string;
+  groupLabelProps?: React.HTMLAttributes<HTMLLIElement>;
+  label: ContentNode;
+}
+
+const ListboxGroup = forwardRef<HTMLElement, ListboxGroupProps>(
+  (
+    {
+      as: Component = 'ul',
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/ListboxGroup.tsx:14]: I wonder what if a user changes `as` into a `div`? The we will have `li` under `div`?

File: packages/react/src/components/Listbox/ListboxOption.tsx [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    useEffect(() => {
+      const element = listboxOptionRef.current;
+
+      setOptions((options) => {
+        const option = { element, value: optionValue };
+        // istanbul ignore next
+        if (!element) return options;
+
+        // Elements are frequently appended, so check to see if the newly rendered
+        // element follows the last element element first before any other checks
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/ListboxOption.tsx:56]: `last element first`?

File: packages/react/src/components/Listbox/ListboxContext.tsx (hunk 9 above)
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxContext.tsx:29]: What do you mean?

File: packages/react/src/components/Listbox/ListboxGroup.tsx (hunk 10 above)
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxGroup.tsx:14]: Correct. We don't have control over if the author changes the element to something that might not be valid markup, we just provide the hook to do so.

File: packages/react/src/components/Listbox/ListboxOption.tsx [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+        // istanbul ignore next
+        // this should never happen, but just in case fall back to options
+        return options;
+      });
+
+      return () => {
+        setOptions((opts) => opts.filter((opt) => opt.element !== element));
+      };
+    }, [optionValue]);
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/ListboxOption.tsx:86]: Took me some time to understand this block, but I think I get it now. It handles appending `listOptions` at the end of the options or in the middle of it.

File: packages/react/src/components/Listbox/ListboxContext.tsx [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  selected: T | null;
+  setOptions: React.Dispatch<React.SetStateAction<T[]>>;
+  onSelect: (option: T) => void;
+};
+
+type ListboxProvider<T extends ListboxOption> = {
+  children: React.ReactNode;
+} & ListboxContext<T>;
+
+/* istanbul ignore next */
```
Comment (dequejosie) [inline packages/react/src/components/Listbox/ListboxContext.tsx:22]: > Why is typescript not happy here?
Sorry, I meant here.

File: packages/react/src/components/Listbox/Listbox.tsx (hunk 5 above)
Comment (scurker) [reply packages/react/src/components/Listbox/Listbox.tsx:98]: `onSelect` happens before the selection happens. `onSelectionChange` happens only after the the change has actually happened.

I suppose we could simplify this by only having a single event that responds when a selection changes, but only actually set the selection when uncontrolled:

```tsx
const handleSelectionChange = useCallback(
  (option: ListboxOption) => {
    onSelectionChange?.({ 
      target: option.element, 
      value: option.value as string 
      previousValue: selectedOption?.value
    });

    if (!isControlled) {
      setActiveOption(option);
      setSelectedOption(option);
    }
  },
  [isControlled, selectedOption]
);
```

Would that help make it more clear?

File: packages/react/src/components/Listbox/ListboxContext.tsx (hunk 13 above)
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxContext.tsx:22]: I'm still not clear what you're referring to. I do not have any typescript errors:

vscode showing no errors on the referenced line

File: packages/react/src/components/Listbox/ListboxOption.tsx (hunk 12 above)
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxOption.tsx:86]: Yeah, it's a bit tricky. We need to pass the rendered options back up the context but need to sort the list based on dom order here.

I went this route because passing down props via a set of unknown components is gnarly, and this abstracts that away completely.

File: packages/react/src/components/Listbox/Listbox.tsx (hunk 5 above)
Comment (thuey) [reply packages/react/src/components/Listbox/Listbox.tsx:103]: Personally, I like the suggested alternative. I do feel like it would make it clearer

File: packages/react/src/components/Listbox/ListboxOption.tsx (hunk 12 above)
Comment (dequejosie) [reply packages/react/src/components/Listbox/ListboxOption.tsx:86]: When I saw the context, I also compared it with passing down the props in my mind and I definitely think context is better.

File: packages/react/src/components/Listbox/ListboxContext.tsx (hunk 13 above)
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxContext.tsx:22]: nyc was complaining about none of the options getting invoked but I don't think we really care about these lines for coverage since it's stubbing out an empty context so I ignored these lines. ¯\\\_(ツ)_/¯

File: packages/react/src/components/Listbox/Listbox.tsx (hunk 5 above)
Comment (scurker) [reply packages/react/src/components/Listbox/Listbox.tsx]: I just pushed that change.

File: packages/react/src/components/Listbox/ListboxOption.tsx [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+          isElementPreceding(
+            options[options.length - 1].element,
+            option.element
+          )
+        ) {
+          return [...options, option];
+        }
+
+        for (const opt of options) {
+          if (isElementPreceding(element, opt.element)) {
```
Comment (thuey) [inline packages/react/src/components/Listbox/ListboxOption.tsx:68]: It seems like this might not work properly when elements are added. I did some testing and `Node.DOCUMENT_POSITION_PRECEDING` is not the same as "immediately preceding", but this code seems to assume that's how that works.

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+          return;
+        }
+
+        const [up, down, home, end, enter, space] = keys;
+        const firstOption = enabledOptions[0];
+
+        if (!activeOption) {
+          setActiveOption(firstOption);
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (scurker) [reply packages/react/src/components/Listbox/Listbox.tsx:129]: Focusing the component should set the first option as active for listbox, but that's not always necessarily the case for composed components that would set the active descendent and need to manage state in some other way (such as combobox).

File: packages/react/src/components/Listbox/ListboxContext.tsx
[the remark this answers is not in the document]
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxContext.tsx:1]: Smaller files / separation of concerns?? If it's just a nit I'd like to leave this as is.

File: packages/react/src/components/Listbox/Listbox.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (thuey) [reply packages/react/src/components/Listbox/Listbox.tsx:129]: Gotcha. I think the question is still applicable with regards to what the `space` or `enter` behavior should be, though--should it also select the first option?

File: packages/react/src/components/Listbox/ListboxOption.tsx [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    const listboxOptionRef = useSharedRef<HTMLElement>(ref);
+    const [id] = propId ? [propId] : useId(1, 'listbox-option');
+    const isActive =
+      active !== null && active.element === listboxOptionRef.current;
+    const isSelected =
+      selected !== null && selected.element === listboxOptionRef.current;
+    const optionValue =
+      typeof value !== 'undefined'
+        ? value
+        : listboxOptionRef.current?.innerText;
```
[the remark this answers is not in the document]
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxOption.tsx:45]: My thinking here was to mirror what native options do by reflecting the label to the text value when the value is omitted. An empty value/label is also valid, so making it required adds verbosity that I don't think offers much. So I'm opposed to changing this.

I did a bunch of testing around this and the value always gets updated due to the `useEffect` callback that sets options. The value does initially start out as `undefined` but gets updated when the hook re-renders as a side effect. That would be an issue if we removed the use effect, but that also fundamentally changes how the component works.

File: packages/react/src/components/Listbox/ListboxOption.tsx (hunk 14 above)
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxOption.tsx:68]: Correct, it's not immediately proceeding.

From testing this was always triggered with a single node at a time (even when mounting/changing multiple nodes). Even when nodes got sorted, the options gets removed from the list and re-added in order so I didn't encounter issues from testing.

File: packages/react/src/components/Listbox/Listbox.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (scurker) [reply packages/react/src/components/Listbox/Listbox.tsx:129]: No, I don't think it should select the first option.

File: packages/react/src/components/Listbox/ListboxOption.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (thuey) [reply packages/react/src/components/Listbox/ListboxOption.tsx:45]: That's fair. I was also concerned about multiple options having an empty value, but I suppose that could still happen even if you made `value` required without some kind of validation that the values were unique.

tldr; I think this is fine as is.

File: packages/react/src/components/Listbox/ListboxOption.tsx (hunk 14 above)
Comment (thuey) [reply packages/react/src/components/Listbox/ListboxOption.tsx:68]: Even when dynamically adding a new option to the end of the list after the others had already rendered?

**Review by thuey — APPROVED** [review]
Assuming https://github.com/dequelabs/cauldron/pull/1167#discussion_r1304729388 is okay, this LGTM. Nice job! This is pretty cool.

File: packages/react/src/components/Listbox/ListboxOption.tsx (hunk 14 above)
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxOption.tsx:68]: Yes, that's what lines 57-65 are covering. This is with the assumption that the elements are already in the correct dom order, if the last element in the array precedes the newly inserted element it should get tacked on at the end.
